service/history/decisionTaskHandler.go

```diff
@@ -227,34 +236,39 @@ func (handler *decisionTaskHandlerImpl) handleDecisionRequestCancelActivity(
 		return err
 	}
 	activityID := attr.GetActivityId()
-	actCancelReqEvent, ai, isRunning := handler.mutableState.AddActivityTaskCancelRequestedEvent(
+	actCancelReqEvent, ai, err := handler.mutableState.AddActivityTaskCancelRequestedEvent(
 		handler.decisionTaskCompletedID,
 		activityID,
 		handler.identity,
 	)
-	if !isRunning {
-		handler.mutableState.AddRequestCancelActivityTaskFailedEvent(
+	switch err.(type) {
+	case nil:
+		if ai.StartedID == common.EmptyEventID {
+			// We haven't started the activity yet, we can cancel the activity right away and
+			// schedule a decision task to ensure the workflow makes progress.
+			_, err = handler.mutableState.AddActivityTaskCanceledEvent(
+				ai.ScheduleID,
+				ai.StartedID,
+				actCancelReqEvent.GetEventId(),
+				[]byte(activityCancellationMsgActivityNotStarted),
+				handler.identity,
+			)
+			if err != nil {
+				return err
+			}
+			handler.activityNotStartedCancelled = true
+		}
+		return nil
+	case *workflow.BadRequestError:
+		_, err = handler.mutableState.AddRequestCancelActivityTaskFailedEvent(
 			handler.decisionTaskCompletedID,
 			activityID,
 			activityCancellationMsgActivityIDUnknown,
 		)
```

Comment:
Here the original error is overwritten. So you might want to add some logs if you don't have them inside mutable calls.

Reply:
this is intentional, basically, mutable state will return 2 kinds of errors,
1. internal server error, which means there is something wrong
2. bad request, for failing decisions